admittance_controller/include/admittance_controller/admittance_rule.hpp

```diff
@@ -0,0 +1,224 @@
+// limitations under the License.
+//
+/// \authors: Denis Stogl, Andy Zelenak, Paul Gesel
+
+#ifndef ADMITTANCE_CONTROLLER__ADMITTANCE_RULE_HPP_
+#define ADMITTANCE_CONTROLLER__ADMITTANCE_RULE_HPP_
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "tf2_ros/buffer.h"
+
+#include "admittance_controller_parameters.hpp"
+#include "control_msgs/msg/admittance_controller_state.hpp"
+#include "control_toolbox/filters.hpp"
+#include "controller_interface/controller_interface.hpp"
+#include "geometry_msgs/msg/wrench_stamped.hpp"
+#include "kinematics_interface/kinematics_interface.hpp"
+#include "pluginlib/class_loader.hpp"
+#include "tf2_eigen/tf2_eigen.h"
+#include "tf2_geometry_msgs/tf2_geometry_msgs.hpp"
+#include "tf2_kdl/tf2_kdl/tf2_kdl.hpp"
+#include "tf2_ros/transform_listener.h"
+// TODO(destogl): remove this include - this is wrong!
+#include "trajectory_msgs/msg/detail/joint_trajectory_point__struct.hpp"
+
+namespace admittance_controller
+{
+struct Transforms
+{
+  Eigen::Isometry3d base_control_;
+  Eigen::Isometry3d base_ft_;
+  Eigen::Isometry3d base_tip_;
+  Eigen::Isometry3d world_base_;
+  Eigen::Isometry3d base_sensor_;
+  Eigen::Isometry3d base_cog_;
+};
+
+struct AdmittanceState
+{
+  // TODO(destogl): does this constructor makes sense at all?
+  AdmittanceState() = default;
+
+  explicit AdmittanceState(size_t num_joints)
+  {
+    admittance_velocity.setZero();
+    admittance_acceleration.setZero();
+    damping.setZero();
+    mass.setOnes();
+    mass_inv.setZero();
+    stiffness.setZero();
+    selected_axes.setZero();
+    current_joint_pos = Eigen::VectorXd::Zero(num_joints);
+    joint_pos = Eigen::VectorXd::Zero(num_joints);
+    joint_vel = Eigen::VectorXd::Zero(num_joints);
+    joint_acc = Eigen::VectorXd::Zero(num_joints);
+  }
+
+  Eigen::VectorXd current_joint_pos;
+  Eigen::VectorXd joint_pos;
+  Eigen::VectorXd joint_vel;
+  Eigen::VectorXd joint_acc;
+  Eigen::Matrix<double, 6, 1> damping;
+  Eigen::Matrix<double, 6, 1> mass;
+  Eigen::Matrix<double, 6, 1> mass_inv;
+  Eigen::Matrix<double, 6, 1> selected_axes;
+  Eigen::Matrix<double, 6, 1> stiffness;
+  Eigen::Matrix<double, 6, 1> wrench_base;
+  Eigen::Matrix<double, 6, 1> admittance_acceleration;
+  Eigen::Matrix<double, 6, 1> admittance_velocity;
+  Eigen::Isometry3d admittance_position;
+  Eigen::Matrix<double, 3, 3> rot_base_control;
+  Eigen::Isometry3d ref_trans_base_ft;
+  std::string ft_sensor_frame;
+};
+
+class AdmittanceRule
+{
+public:
+  explicit AdmittanceRule(const std::shared_ptr<rclcpp_lifecycle::LifecycleNode> & lifecycl_node)
+  {
+    parameter_handler_ = std::make_shared<admittance_controller::ParamListener>(
+      lifecycl_node->get_node_parameters_interface());
+    parameters_ = parameter_handler_->get_params();
+    num_joints_ = parameters_.joints.size();
+    admittance_state_ = AdmittanceState(num_joints_);
```

Comment:
Why don't you make this a pure interface and do makes *.cpp file if needed? Nevertheless, this could go into implementation too.